src/libopensc/card-starcos.c

```diff
@@ -280,6 +281,54 @@ static unsigned int starcos_determine_pin_encoding(sc_card_t *card)
 	return encoding;
 }
 
+/**
+ * Clears logged in state for all pins
+ */
+static void starcos_pin_clear_logged_in(sc_card_t *card) {
+	starcos_ex_data * ex_data = (starcos_ex_data*)card->drv_data;
+	memset(&ex_data->pin_logged_in, 0, sizeof(ex_data->pin_logged_in));
+}
+
+/**
+ * Set logged in state for the given pin reference.
+ * Returns -1 if the pin state cannot be set
+ * Note: pin reference must be greater than 0
+ */
+static int starcos_pin_set_logged_in(sc_card_t *card, unsigned char pin_reference) {
+	unsigned int i;
+	starcos_ex_data * ex_data = (starcos_ex_data*)card->drv_data;
+
+	// check if the pin has already been registered
+	for(i=0; i<sizeof(ex_data->pin_logged_in)/sizeof(unsigned char); i++) {
+		if ( ex_data->pin_logged_in[i] == pin_reference ) return i;
+	}
+	// pin not yet seen, register to the first available slot
+	for(i=0; i<sizeof(ex_data->pin_logged_in)/sizeof(unsigned char); i++) {
+		if ( ex_data->pin_logged_in[i] == 0 ) {
+			ex_data->pin_logged_in[i] = pin_reference;
+			return i;
+		}
+	}
+	// could not register pin logged in state
+	return -1;
+}
+
+/**
+ * Returns 1 if the given pin reference is logged in
+ * Otherwise returns 0
+ */
+static int starcos_pin_is_logged_in(sc_card_t *card, unsigned char pin_reference) {
+	unsigned int i;
+	starcos_ex_data * ex_data = (starcos_ex_data*)card->drv_data;
+
+	for(i=0; i<sizeof(ex_data->pin_logged_in)/sizeof(unsigned char); i++) {
+		if ( ex_data->pin_logged_in[i] == pin_reference ) {
+			return 1;
+		}
+	}
+	return 0;
+}
```

Comment:
Yes, but the same "Command Sequence" section of the 3.5 documentation says:
After successful command execution, STARCOS performs the following: 1. Sets the state of the PIN or password to „verified“.
In practice, the empty VERIFY also works on the the v3.5 cards, but it never answers '90 00', but always 'C3 Cx'. This means in Firefox/P11 the login function will not work. 
That's why this workaround is only used for v3.5 cards, but I agree, it would be better to determine the logged_in status by the card, I will double check if it is possible.

Reply:
Some PIV like cards have the same problem with "empty VERIFY" and testing for login state. 
`#define CI_VERIFY_LC0_FAIL                  0x00000002U /* VERIFY Lc=0 never returns 90 00 if PIN not needed */` and addresses it by `/* try another method, looking at a protected object this may require adding one of these to NEO */`  Look at the logic starting at line 5281 ` /* Some cards never return 90 00  for SC_PIN_CMD_GET_INFO even if the card state is verified */`.

Reply:
Unfortunately I did not find any StarCOS 3.x data object or file which could produce the "69 82"/"Security status not satisfied" error besides using a private key. But the card driver itself is unaware of any private keys, this information comes from the higher layers like PKCS15. I keep searching for a solution

Reply:
I was stuck with this issue a little bit :) 
At the end, I even managed to get a contact from the card manufacturer, but unfortunately they have confirmed: it is not possible to query the password's security status on the card with this Firmware Version of StarCOS 3.5 (IFAIK v3.52). Should I leave or remove the software workaround?